doc/neps/nep-0054-simd-cpp-highway.rst

```diff
@@ -0,0 +1,480 @@
+- pulling in individual bits or algorithms of Highway (as in `gh-24018`_) or
+  SLEEF (as discussed in that same PR)
+
+
+Usage and Impact
+----------------
+
+N/A - there will be no significant user-visible changes.
+
+
+Backward compatibility
+----------------------
+
+There will be no changes in user-facing Python or C APIs: all the methods to
+control compilation and runtime CPU feature selection should remain, although
+there may be some changes due to moving to Meson and C++ without regards to the
+Highway/Universal Intrinsics choice.
+
+On Windows, MSVC may have to be avoided, as a result of Highway's use of
+pragmas which are less well supported by MSVC. This means that we likely have
+to build our wheels with clang-cl or Mingw-w64. Both of those should work - we
+merged clang-cl support a while back (see `gh-20866`_), and SciPy builds with
+Mingw-w64. It may however impact other redistributors or end users who build
+from source on Windows.
+
+The potential impact on licensing because of Highway's use of the Apache 2.0
+license needs discussing before Highway can be accepted. Or Highway would have
+to have to dual-license under Apache 2.0 and an MIT/BSD type license.
+
+
+High-level considerations
+-------------------------
+
+.. note::
+
+   Currently this section attempts to cover each topic separately, and
+   comparing the future use of a NumPy-specific C++ implementation vs. use of
+   Google Highway with our own numerical routines on top of that. It does not
+   (yet) assume a decision or proposed decision is made. Hence this NEP is not
+   "this is proposed" with another option in the Alternatives section, but
+   rather a side-by-side comparison.
+
+    
+Development effort and long-term maintainability
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Moving to Highway is likely to be a significant development effort.
+Longer-term, this will hopefully be offset by Highway itself having more
+maintainer bandwidth to deal with ongoing issues in compiler support and adding
+new platforms. It should increase the bus factor of SIMD support in NumPy to
+some extent. That said - the Highway repository also seems to show that its bus
+factor is low - it was 1 until the start of 2023, and now it's perhaps 2.
+
+Highway being used by other projects, like Chromium and `JPEG XL`_ (see
+`this more complete list <https://google.github.io/highway/en/master/README.html#examples>`__
+in the Highway documentation), does imply that there is likely to be a benefit
+of a wider range of testing and bug reporting/fixing.
+
+One concern is that new instructions may have to be added, and that that is
+often best done as part of the process of developing the numerical kernel that
+needs the instruction. This will be a little more clumsy if the instruction
+lives in Highway which is a git submodule inside the NumPy repo - there will be
+a need to implement a temporary/generic version first, and then update the
```

Comment:
I think concern here is relative to a solution where NumPy owns and maintains its own C++ universal intrinsic. Adding new optimizations which requires new intrinsics (or some combination of) will be a lot easier if the entire patch is in one PR and reviewed once in one place. Making changes to a git submodule means having to go through two separate pull requests which will slow things down and makes developing a new feature a tad bit harder. I should also point out that benchmarking perf with a patch that makes changes to submodule is a pain point too. You either have to upstream your changes to submodule first before benchmarking (which isn't always possible) or jump through git submodule hoops to ensure asv builds your branch with the changes in submodule.

Reply:
Can you elaborate on why that's not possible by simply extending Highway temporarily within NumPy? You seem to be making the assumption that we will always have to update within Highway first, whilst another approach would be to build the intrinsic for NumPy's use case and then attempt to upstream it. This mirrors the development process which took place with `npyv_negative_@sfx@` where it appeared in `tanh` first and then got ported to a more central place later (what we have today).

Any improvements in Highway to adopt a new intrinsic would then allow us to delete the code, but we still retain a straight forward development flow from a NumPy point of view.

Reply:
> Can you elaborate on why that's not possible by simply extending Highway temporarily within NumPy? 

You absolutely could, but I think we would want to keep that to a minimum. BTW, going through two pull requests isn't a big deal either. I mentioned it because it does change the workflow from a developer point of view.

Reply:
I don't understand? Why would we want to change the workflow when we can keep it similar to how it is now?

Reply:
Just to be explicit about the proposed workflow. Because adding ops in Highway is just a matter of defining a new function in namespace hwy::HWY_NAMESPACE, this can be done in numpy code `#ifndef HWY_OP_NEGATIVE` (or whatever the op is called). So you'd put the same implementation that will go into Highway, inside numpy for purposes of testing/benchmarking, which you can thus do in a single pull request without delay. And when Highway adopts this new op, it defines `HWY_OP_NEGATIVE` so there is no conflict.

Reply:
I am going to resolve this discussion, as I think the text adequately reflects the slight inconvenience of maintaining a forked subrepo.